src/Features/LanguageServer/Protocol/Workspaces/LspWorkspaceManager.cs

```diff
@@ -283,23 +315,40 @@ async Task<(Solution Solution, bool IsForked)> GetLspSolutionForWorkspaceAsync(W
                 return (workspaceCurrentSolution, IsForked: false);
             }
 
-            // Step 3: See if we can reuse a previously forked solution.
+            // Step 4: See if we can reuse a previously forked solution.
             if (cachedSolution != default && cachedSolution.forkedFromVersion == workspaceCurrentSolution.WorkspaceVersion)
-            {
                 return (cachedSolution.solution, IsForked: true);
-            }
 
-            // Step 4: Fork a new solution from the workspace with the LSP text applied.
+            // Step 5: Fork a new solution from the workspace with the LSP text applied.
             var lspSolution = workspaceCurrentSolution;
             foreach (var (uri, workspaceDocuments) in documentsInWorkspace)
-            {
                 lspSolution = lspSolution.WithDocumentText(workspaceDocuments.Select(d => d.Id), _trackedDocuments[uri]);
-            }
 
             // Remember this forked solution and the workspace version it was forked from.
             _cachedLspSolutions[workspace] = (workspaceCurrentSolution.WorkspaceVersion, lspSolution);
             return (lspSolution, IsForked: true);
         }
+
+        void OpenOrEditDocumentsInMutatingWorkspace(Workspace workspace)
+        {
+            if (workspace is not ILspWorkspace { SupportsMutation: true } mutatingWorkspace)
+                return;
+
+            var currentSolution = workspace.CurrentSolution;
+            foreach (var (uri, sourceText) in _trackedDocuments)
+            {
+                foreach (var documentId in currentSolution.GetDocumentIds(uri))
+                {
+                    // Note: there is a race here in that we might see/change/return here based on the
+                    // relationship of 'sourceText' and 'currentSolution' while some other entity outside of the
+                    // confines of lsp queue might update the workspace externally.  That's completely fine
+                    // though.  The caller will always grab the 'current solution' again off of the workspace
+                    // and check the checksums of all documents against the ones this workspace manager is
+                    // tracking.  If there are any differences, it will fork and use that fork.
+                    mutatingWorkspace.UpdateTextIfPresent(documentId, sourceText);
```

Comment:
Have we verified this all no-ops if this is repeatedly being called in a loop? It'd be unfortunate if on every keystroke we end up creating a bunch of new solution objects for text changes we've already applied.

Reply:
Yes.  I debugged through this many times, and it was wonderful seeing it no-op out of actually forking :)